Add todo component

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add todo component. Updates the Shopping List component to implement the todo platform. This platform was also manually tested with a local todo implementation (removed as it is a larger scope PR, so local todo list based on rfc5545 will come in followups).

The to-do list platform registers the existing shopping list frontend panel (can be updated based on frontend PR direction)

An alternative can be to import shopping list into a local todo list component, though we can probably do that in followup PRs if desired.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to architecture proposal: https://github.com/home-assistant/architecture/discussions/962
- Link to core pull request: https://github.com/home-assistant/core/pull/100019
- ~~Link to frontend pull request: https://github.com/home-assistant/frontend/pull/17883~~
- Link to frontend pull request: https://github.com/home-assistant/frontend/pull/18289
- Link to brands pull request: https://github.com/home-assistant/brands/pull/4682
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28925
- Link to developer documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/1929

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

allenporter committed:
Add todo component

joostlek commented on homeassistant/components/demo/todo.py:
```diff
@@ -0,0 +1,35 @@
+
+
+class DemoTodoListEntity(TodoListEntity):
+    """Representation of a demo todo entity."""
```
```suggestion
    """Representation of a demo todo list entity."""
```

allenporter commented:
Marking as draft until we complete a round of user research to make sure the proposal has all the requirements. I think the APIs are still mostly correct here, but marking to indicate the actual status.

allenporter committed:
Fix lint error and revert frontend panel registration

MartinHjelmare commented on homeassistant/components/demo/__init__.py:
```diff
@@ -43,6 +43,7 @@
     Platform.SIREN,
     Platform.STT,
     Platform.SWITCH,
+    Platform.TODO,
```
Please sort 🔡.

MartinHjelmare commented on homeassistant/components/shopping_list/__init__.py:
```diff
@@ -209,18 +227,33 @@ async def async_add(self, name, context=None):
-            raise NoMatchingShoppingListItem
-
-        self.items.remove(item)
+    async def async_remove_items(self, item_ids: set[str], context=None):
```
```suggestion
    async def async_remove_items(self, item_ids: set[str], context=None) -> None:
```

MartinHjelmare commented on homeassistant/components/shopping_list/__init__.py:
```diff
@@ -287,16 +323,46 @@ def async_reorder(self, item_ids, context=None):
             context=context,
         )
 
+    async def async_move_item(self, uid: str, previous: str | None = None):
```
```suggestion
    async def async_move_item(self, uid: str, previous: str | None = None) -> None:
```

MartinHjelmare commented on homeassistant/components/shopping_list/__init__.py:
```diff
@@ -180,9 +183,22 @@ async def sort_list_service(call: ServiceCall) -> None:
+    if unload_ok := await hass.config_entries.async_unload_platforms(
+        config_entry, PLATFORMS
+    ):
+        hass.data.pop(DOMAIN)
```
The services and websocket command will fail if we remove the shopping list data.

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -0,0 +1,310 @@
+        """Add an item to the To-do list."""
+        raise NotImplementedError()
+
+    async def async_delete_todo_items(self, uids: set[str]) -> None:
```
Do we use a set parameter with multiple uids to follow the rfc spec or why do we use multiple instead of a single item for this action?

allenporter replied:
This is purely for practical performance concerns, not related to the rfc spec.

When reviewing the shopping list API, one observation was that we currently require a "clear completed items" for the shopping list.  This is also part of the new spec for todo when I looked at the wireframes shared by Bram that Madelena made.  (This is mentioned in https://github.com/home-assistant/architecture/discussions/962 in passing under "Consequences" as a performance topic)

An alternative could be to rename this "clear completed items" rather than supporting delete of arbitrary items, however then i believe it requires more custom logic by the integration to implement. I liked simplifying this by pushing this to be a frontend concern about which items it is deleting (delete is needed anyway) allowing cloud integrations to have a more straight forward API. 

Open to suggestions.

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -0,0 +1,310 @@
+        raise NotImplementedError()
+
+    async def async_update_todo_item(self, item: TodoItem) -> None:
+        """Add an item to the To-do list."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -0,0 +1,310 @@
+        raise NotImplementedError()
+
+    async def async_delete_todo_items(self, uids: set[str]) -> None:
+        """Add an item to the To-do list."""
```
Stale docstring.

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -0,0 +1,310 @@
+    """Handle the list of To-do items in a To-do- list."""
+    component: EntityComponent[TodoListEntity] = hass.data[DOMAIN]
+    if (
+        not (entity_id := msg.get(CONF_ENTITY_ID))
```
The entity id isn't optional.

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -0,0 +1,310 @@
+        return
+
+    try:
+        await entity.async_create_todo_item(item=TodoItem(**msg["item"]))
```
We pin the frontend version to the core version but maybe still nice to be explicit and name the keyword arguments?

MartinHjelmare commented on tests/components/todo/test_init.py:
```diff
@@ -0,0 +1,146 @@
+        yield
+
+
+async def test_list_todo_items(
```
We should add tests for the other actions too and the entity state.

allenporter replied:
I have added tests for the state.

The other actions are tested below, but just showing that they are unsupported by demo. When we look at the coverage for todo alone its not great:
```
$ py.test tests/components/todo --cov=homeassistant.components.todo --cov-report=term-missing
Name                                        Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------
homeassistant/components/todo/__init__.py     128     22    83%   85, 175-183, 219-224, 265-270, 304-309
homeassistant/components/todo/const.py         10      0   100%
-------------------------------------------------------------------------
TOTAL                                         138     22    84%
```

However, the actions are also tested by shopping list tests which brings the coverage to 96%:
```
$ py.test tests/components/todo tests/components/shopping_list --cov=homeassistant.components.todo --cov-report=term-missing
Name                                        Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------
homeassistant/components/todo/__init__.py     128      5    96%   85, 180-181, 221-222
homeassistant/components/todo/const.py         10      0   100%
-------------------------------------------------------------------------
TOTAL                                         138      5    96%
```

Where the only failures are when `HomeAssistatError` is raised which doesn't happen (but will with the cloud APIs which we can also exercise)

I can add support to demo but i think it would replicate the functionality in shopping list or local todo -- perhaps you have a different testing approach in mind  (`patch` out the demo entity calls to succeed or raise?).

Given the above, let me know if that is sufficient or if you think there is still additional value we can get from more tests happy to add them if you can elaborate a bit

MartinHjelmare replied:
We should not rely on the demo or shopping list tests to cover the todo integration.

Here's how we test the event entity, with a mock platform in at least one test.

https://github.com/home-assistant/core/blob/dev/tests/components/event/test_init.py

allenporter replied:
Thanks, i removed the demo changes and added mock platforms here.

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -0,0 +1,310 @@
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+
+WEBSOCKET_TODO_ITEM_CREATE_SCHEMA = vol.Schema(
```
Why don't we inline these schemas?

allenporter committed:
Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Address PR feedback

allenporter committed:
Remove shopping list config entry unload support and fix typing

allenporter committed:
Add tests for state

allenporter committed:
Use mock platform for todo tests

allenporter committed:
Revert demo todo platform changes

allenporter committed:
Add test coverage for exception cases

allenporter committed:
Add integration title

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -0,0 +1,312 @@
+    @property
+    def todo_items(self) -> list[TodoItem] | None:
+        """Return the To-do items in the To-do list."""
+        if hasattr(self, "_attr_todo_items"):
```
This will always be true. Just return `self._attr_todo_items`.

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -0,0 +1,312 @@
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_create)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_delete)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_update)
+    websocket_api.async_register_command(hass, websocket_handle_todo_item_move)
```
Shouldn't we also register services, besides websocket commands, for all the actions?

MartinHjelmare replied:
Or maybe only do services? What's the reason for the websocket command instead of service?

allenporter replied:
> What's the reason for the websocket command instead of service?

I wouldn't say *instead of*, its just limiting scope for initial PR to include what frontend needs (e.g. what  https://github.com/home-assistant/frontend/pull/17883 used though that will be replaced). Happy to change or increase scope here. Part of this may be a misunderstanding (see net). 

> The other four ones,... I would do as services

OK so:
- We prefer reusing services in FE calls whenever possible instead of dedicated websocket calls? My impression was the design standard was that we duplicate websocket APIs and services calls, based on everything I had previously seen, but that may just be inferring somethign incorrect from the legacy code base.   I am definitely all for not duplicating websockets and services, it just has not been a standard I've seen anywhere.
- Given that, What is the distinction on read vs write duplicating functionality in websockets and services? 

Regarding list items as a service: Are we aligned on https://github.com/home-assistant/architecture/discussions/948#discussioncomment-7197201 to move forward yet? 



Aside: One reason for avoiding services was lack of movement on the service call responses discussion (we're still debating details 1+ year out so have been looking to see more alignment there). Since I first created this in the mean time there has been more discussion, though would still like to see a decision to move forward and next steps are opaque.

MartinHjelmare replied:
The actions that change state don't return anything currently so I don't think the frontend needs a websocket command. There's no data to return and no subscription needed. A service call should be fine for those.

For the list items action we need to return data. The frontend normally wants a websocket command in that case is my experience. A service would be nice to leverage in automations but may not be needed initially.

The frontend team should confirm this.

allenporter replied:
Thank you for the detail, got it. I've updated this to use services for the mutating actions, preserving the list websocket. I confirmed with another opinion that services are OK.  All tests were rewritten to use the services instead of websockets.

We'll need to add the list service with responses before the services will be useful outside the frontend, but that should be feasible to do.

allenporter replied:
(I'll want to add service response values for the create item service as well)

allenporter replied:
These action apis were made for frontend but I'm wondering if they'll be usable by end users.

Shopping list was designed for matching tasks by name, but that is not how the various cloud task apis work. However, given home assistant make assumptions about finite task list limits, we could still make it work by name without ids. As in, delete takes names instead of IDs.


(Relatedly ,we could enforce the limits in the entity e.g. no more than 500 items in the Todo list etc)

MartinHjelmare replied:
Sounds worth to explore.

allenporter committed:
Remove unnecessary if check

allenporter committed:
Move from WS to services and fix enum types

allenporter committed:
Readability improvements

allenporter committed:
Fix debugging and comment issues

allenporter committed:
Rewrite service calls as external functions

allenporter committed:
Move services to work by name

allenporter committed:
Readability improvement

allenporter reviewed (commented):
I took another pass on the service APIs given an offline discussion about making this easier to use in automations, supporting update, delete, move by "summary". The APIs now accept either a uid or summary and will use the current todo item state to retrieve.

There may be additional test simplification we can do like streamlining fixtures, but I wanted to get earlier input on the service API direction. Please have a look!

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -57,24 +57,45 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+            cv.has_at_most_one_key("previous_uid", "previous_summary"),
+            # Don't mix and match uid and summary across current and previous
+            cv.has_at_most_one_key("uid", "previous_summary"),
+            cv.has_at_most_one_key("summary", "previous_uid"),
```
It's easier to read if we always start with the same pair type.
```suggestion
            cv.has_at_most_one_key("previous_uid", "summary"),
```

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -185,21 +214,45 @@ async def _async_create_todo_item(entity: TodoListEntity, call: ServiceCall) ->
+    if not item.uid:
+        found = _find_by_summary(item.summary, entity.todo_items)
+        if not found:
+            raise vol.Invalid(
```
We're no longer in the service validation, but in the service handler here. Then we should raise `ValueError` for bad user input and `HomeAssistantError` for api call failures and similar errors. `vol.Invalid` should only be raised from within a `vol.Schema` validation.

allenporter replied:
I remember we have an arch discussion to start on this one... Updated!

MartinHjelmare commented on homeassistant/components/todo/strings.json:
```diff
@@ -40,25 +40,37 @@
-        "previous": {
+        "summary": {
+          "name": "Summary",
+          "description": "The summary of the To-dot represents the To-do item."
```
```suggestion
          "description": "The summary of the To-do represents the To-do item."
```

MartinHjelmare commented on homeassistant/components/todo/strings.json:
```diff
@@ -22,7 +22,7 @@
     "update_item": {
       "name": "Update To-do List Item",
-      "description": "Update an existing To-do List Item.",
+      "description": "Update an existing To-do List Item based on either it's Unique Id or Summary.",
```
```suggestion
      "description": "Update an existing To-do List Item based on either its Unique Id or Summary.",
```

MartinHjelmare commented on homeassistant/components/todo/strings.json:
```diff
@@ -40,25 +40,37 @@
     "delete_item": {
       "name": "Delete a To-do List Item",
-      "description": "Delete an existing To-do List Item.",
+      "description": "Delete an existing To-do List Item either by it's Unique Id or Summary.",
```
```suggestion
      "description": "Delete an existing To-do List Item either by its Unique Id or Summary.",
```

MartinHjelmare commented on homeassistant/components/todo/strings.json:
```diff
@@ -40,25 +40,37 @@
     "move_item": {
       "name": "Re-order a To-do List Item",
-      "description": "Re-orde an existing To-do List Item within the To-do List.",
+      "description": "Re-orde an existing To-do List Item within the To-do List either by it's Unique Id or Summary. The item is either moved to the first position in the list or immediately following the item identified by the Previous Unique Id or Summary.",
```
```suggestion
      "description": "Re-orde an existing To-do List Item within the To-do List either by its Unique Id or Summary. The item is either moved to the first position in the list or immediately following the item identified by the Previous Unique Id or Summary.",
```

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -30,7 +30,7 @@
 
 TASK_ITEM_FIELDS = {
-    vol.Required("summary"): cv.string,
+    vol.Required("summary"): vol.All(cv.string, vol.Length(min=1)),
```
`summary` is still required in the schema. Should that change?

allenporter replied:
This comment got me thinking about partial update in general.

I've gone ahead and pushed the schemas back into the service directly so we can diverge between create and update if needed. Now they diverge.

For Create, i think it makes sense to stay required.
For Update, the difference is about partial update. I *think* users would want partial update support to make it easier to just change the fields we want to change, without having to read the existing item first. As a result, I went ahead and made this optional and adjusted the test cases. Here are the cases:

todo.update_item:   # Update the summary only, leave status alone
  - uid: 12345
    summary: "foo"

todo.update_item:   # Update the status only, leave summary alone
  - uid: 12345
    status: "completed"


This required updating the data classes parsing and tyying (now slightly less useful since everything is optional across create/udpate cases).

As a consequence, this is now possible, but a no-op, though I think its ok to leave as is.

todo.update_item:
  - summary: "foo"

MartinHjelmare commented on homeassistant/components/todo/services.yaml:
```diff
@@ -0,0 +1,84 @@
+        - todo.TodoListEntityFeature.MOVE_TODO_ITEM
+  fields:
+    uid:
+      required: true
```
All these parameters are optional now.

allenporter replied:
Corrected, then I went ahead and removed all the `required: false` given the default is already false.

allenporter committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Raise ValueError instead of vol.Invalid

allenporter committed:
Add todo update_item schema validation tests

allenporter committed:
Update todo services.yaml required fields

allenporter committed:
Partial todo item updates

allenporter committed:
Update shopping list partial updates

MartinHjelmare commented on homeassistant/components/todo/__init__.py:
```diff
@@ -233,7 +235,7 @@ async def _async_delete_todo_items(entity: TodoListEntity, call: ServiceCall) ->
             item = _find_by_summary(summary, entity.todo_items)
             if not item:
-                raise vol.Invalid(f"Unable to find To-do item with summariy'{summary}")
+                raise ValueError(f"Unable to find To-do item with summariy'{summary}")
```
```suggestion
                raise ValueError(f"Unable to find To-do item with summary '{summary}")
```

MartinHjelmare commented on homeassistant/components/todo/services.yaml:
```diff
@@ -27,16 +26,14 @@ update_item:
         text:
     summary:
-      required: false
+      required: true
```
Summary isn't required for update.

allenporter replied:
Removed.

allenporter committed:
Update homeassistant/components/todo/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Remove required field

allenporter committed:
Revert move item service, back to websocket

allenporter committed:
Change move API "previous_uid" to "pos" to simplify frontend

allenporter committed:
Remove stale comment

allenporter committed:
Improve readability reordering websocket

allenporter reviewed (commented):
Made two API changes based on offline discussion for FE concerns:
- The move service was reverted back to a websocket to be frontend focused (also lower priority for end user service)
- The move service was simplified to take a position rather than a previous item. This pushes complexity into integrations that need to handle relative positions rather than in the frontend.

MartinHjelmare reviewed (approved):
Looks good!

Is codecov correct that some lines are not covered? Can we test those in that case?

allenporter committed:
Improve todo component test converage

allenporter commented:
> Looks good!
> 
> Is codecov correct that some lines are not covered? Can we test those in that case?

Looks legit. I updated todo component with additional tests.

The shopping list case for unloaded is not tested, but the code is still there. One option is to just remove this code since shopping list does not yet support unload, though it feels weird to leave it incomplete.

MartinHjelmare commented:
I don't see the code regarding unload in the shopping list integration.

allenporter committed:
Remove unnecessary debug statements

allenporter committed:
Remove unload tests since they didn't increase coverage

allenporter commented:
> I don't see the code regarding unload in the shopping list integration.

There is code in `shopping_list/todo.py`:

```
    async def async_added_to_hass(self) -> None:
        """Entity has been added to hass."""
        self.async_on_remove(self._data.async_add_listener(self.async_write_ha_state))
```

And so the add listener unsub code in shopping list turned out to be not actually used given shopping list todo entities cannot be unloaded.

(I thought the code I just added added coverage in todo, but it didn't so i reverted it. the code coverage is similar to event now at least)

MartinHjelmare commented on homeassistant/components/shopping_list/todo.py:
```diff
@@ -0,0 +1,103 @@
+
+    async def async_added_to_hass(self) -> None:
+        """Entity has been added to hass."""
+        self.async_on_remove(self._data.async_add_listener(self.async_write_ha_state))
```
Add a comment here that currently the shopping list integration doesn't support config entry unload.

MartinHjelmare reviewed (changes requested):
We should add back the unload tests for todo integration. Add a mock entity that should be removed or unavailable when the config entry is unloaded. That way we know that we've hooked up the todo integration to the entity component helper correctly.

Example:
https://github.com/home-assistant/core/blob/ab29c796da44392eb61527e7e8c8e976d5596869/tests/components/lawn_mower/test_init.py#L62-L126

allenporter committed:
Improve test coverage for unload

allenporter committed:
Add comment about shopping list config entry unload

allenporter commented:
Thank you, I was missing the call to `async_unload_platforms`.

bramkragten committed:
Move frontend panel from shopping list to todo
